src/common/ui/setting-text.vue

```diff
@@ -23,39 +32,70 @@ export default {
       type: Boolean,
       default: true,
     },
+    hasSave: {
+      type: Boolean,
+      default: true,
+    },
+    hasReset: Boolean,
   },
   data() {
     return {
       value: null,
       jsonValue: null,
       error: null,
+      canSave: null,
+      canReset: null,
     };
   },
   created() {
     const handle = this.json
       ? (value => JSON.stringify(value, null, '  '))
       // XXX compatible with old data format
       : (value => (Array.isArray(value) ? value.join('\n') : value || ''));
-    this.revoke = hookSetting(this.name, val => { this.value = handle(val); });
-    if (this.json) this.$watch('value', this.parseJson);
+    this.revoke = hookSetting(this.name, val => {
+      this.savedValue = val;
+      const text = handle(val);
+      if (this.value !== text) this.value = text;
+      else this.onInput(val);
+    });
+    this.defaultValue = objectGet(defaults, this.name);
   },
   beforeDestroy() {
     this.revoke();
   },
+  watch: {
+    value(val) {
+      if (this.json) {
+        try {
+          val = JSON.parse(val);
+          this.jsonValue = val;
+          this.error = null;
+        } catch (e) {
+          this.error = e.message || e;
+        }
+      }
+      this.onInput(val);
+    },
+  },
   methods: {
-    parseJson() {
-      try {
-        this.jsonValue = JSON.parse(this.value);
-        this.error = null;
-      } catch (e) {
-        this.error = e.message || e;
+    /** `event` is absent when calling from save() */
+    onChange(event) {
+      if (!this.error) {
+        const value = this.json ? this.jsonValue : this.value;
+        if (this.sync || !event) options.set(this.name, value);
```

Comment:
Is this line virtually `onSave`?

Reply:
onSave does nothing except sending an event to the parent, which is why it calls onChange to actually write the value.